automata/parse/postfix.py

```diff
@@ -0,0 +1,132 @@
+            isinstance(curr_token, InfixOperator) or isinstance(curr_token, PostfixOperator)
+        ):
+            raise exceptions.InvalidRegexError(f"Token '{curr_token}' cannot appear at the start of a statement.")
+
+        elif isinstance(prev_token, InfixOperator) and (
+            isinstance(curr_token, InfixOperator) or isinstance(curr_token, PostfixOperator)
+        ):
+            raise exceptions.InvalidRegexError(f"'{prev_token}' cannot appear next to '{curr_token}'.")
+
+        elif isinstance(prev_token, InfixOperator):
+            if curr_token is None:
+                raise exceptions.InvalidRegexError(f"'{prev_token}' cannot appear at the end of a statement.")
+            elif isinstance(curr_token, RightParen):
+                raise exceptions.InvalidRegexError(f"'{prev_token}' cannot appear immediately before ')'.")
+
+        elif isinstance(prev_token, LeftParen):
+            if isinstance(curr_token, RightParen):
+                raise exceptions.InvalidRegexError("Cannot have right paren immediately after left paren.")
+            elif isinstance(curr_token, InfixOperator) or isinstance(curr_token, PostfixOperator):
+                raise exceptions.InvalidRegexError(f"'(' cannot appear immediately before '{prev_token}'.")
+
+            paren_counter += 1
+
+        elif isinstance(prev_token, RightParen):
+            paren_counter -= 1
+
+            if paren_counter < 0:
+                raise exceptions.InvalidRegexError("Token list has mismatched parethesis.")
+
+    if paren_counter != 0:
+        raise exceptions.InvalidRegexError("Token list has unclosed parethesis.")
+
+
+
+def tokens_to_postfix(tokens: List[Token[ExpressionResultT]]) -> List[Token[ExpressionResultT]]:
+    "Takes in tokens and changes them to postfix ordering"
+    stk: Deque[Token] = deque()
+    res: List[Token] = []
+
+    def comp_precedence(a: Token, b: Token) -> bool:
+        "Compare precedence of operators"
+        return a.get_precedence() <= b.get_precedence()
+
+    for c in tokens:
+        if isinstance(c, Literal) or isinstance(c, PostfixOperator):
+            res.append(c)
+        elif isinstance(c, RightParen):
+            while len(stk) > 0 and not isinstance(stk[-1], LeftParen):
+                res.append(stk.pop())
+            stk.pop()
+        elif isinstance(c, LeftParen):
+            stk.append(c)
+        elif not stk or isinstance(stk[-1], LeftParen) or not comp_precedence(c, stk[-1]):
+            stk.append(c)
+        else:
+            while stk and not isinstance(stk[-1], LeftParen) and comp_precedence(c, stk[-1]):
+                res.append(stk.pop())
+            stk.append(c)
+
+    while stk:
+        res.append(stk.pop())
+
+    return res
+
+
+def parse_postfix_tokens(postfix_tokens: List[Token[ExpressionResultT]]) -> ExpressionResultT:
+    """Parse list of postfix tokens to produce value of expression"""
+
+    stack: Deque[ExpressionResultT] = deque()
```

Comment:
@eliotwrobson As much as I personally love deques, is a deque actually faster if the only operations you're performing are `append` (right) and `pop`? I don't see any `appendleft` or `popleft` occurrences in this PR, so was just curious. 😅

Reply:
Based on this blog post, it appears that deques are faster even when only appending and poping from one side: https://dev.to/v_it_aly/python-deque-vs-listwh-25i9

I think it has to do with the way that deques allocate memory (or something like that), that's more efficient than the way lists do it.

Reply:
@eliotwrobson Very well! You have me convinced that deques are the way to go here. 🙂